ext/liquid_c/context.c

```diff
@@ -0,0 +1,128 @@
+
+    // Basic types that do not respond to #evaluate
+    if (klass == rb_cString || klass == rb_cArray || klass == rb_cHash)
+        return expression;
+
+    // Liquid::VariableLookup is by far the most common type after String
+    if (klass == cLiquidVariableLookup)
+        return variable_lookup_evaluate(expression, self);
+
+    if (rb_respond_to(expression, id_evaluate))
+        return rb_funcall(expression, id_evaluate, 1, self);
+
+    return expression;
+}
+
+void context_maybe_raise_undefined_variable(VALUE self, VALUE key)
+{
+    if (rb_ivar_get(self, id_ivar_strict_variables)) {
+        Check_Type(key, T_STRING);
+        rb_enc_raise(utf8_encoding, cLiquidUndefinedVariable, "undefined variable %s", RSTRING_PTR(key));
+    }
+}
+
+VALUE context_find_variable(VALUE self, VALUE key, VALUE raise_on_not_found)
+{
+    VALUE scope = Qnil, variable = Qnil;
+
+    VALUE scopes = rb_ivar_get(self, id_ivar_scopes);
+    Check_Type(scopes, T_ARRAY);
+    const VALUE *scopes_ptr = RARRAY_CONST_PTR(scopes);
+    long scopes_len = RARRAY_LEN(scopes);
+
+    for (long i = 0; i < scopes_len; i++) {
+        VALUE this_scope = scopes_ptr[i];
+        if (TYPE(this_scope) == T_HASH) {
+            // Does not invoke any default value proc, this is equivalent in
+            // cost and semantics to #key? but loads the value as well
+            variable = rb_hash_lookup2(this_scope, key, Qundef);
+            if (variable != Qundef) {
+                scope = this_scope;
+                goto variable_found;
+            }
+        } else if (RTEST(rb_funcall(this_scope, id_has_key, 1, key))) {
+            variable = rb_funcall(this_scope, id_aref, 1, key);
+            goto variable_found;
+        }
+    }
+
+    VALUE environments = rb_ivar_get(self, id_ivar_environments);
+    Check_Type(environments, T_ARRAY);
+    const VALUE *environments_ptr = RARRAY_CONST_PTR(environments);
+    long environments_len = RARRAY_LEN(environments);
+    VALUE strict_variables = rb_ivar_get(self, id_ivar_strict_variables);
+
+    for (long i = 0; i < environments_len; i++) {
+        VALUE this_environ = environments_ptr[i];
+        if (TYPE(this_environ) == T_HASH) {
+            // Does not invoke any default value proc, this is equivalent in
+            // cost and semantics to #key? but loads the value as well
+            variable = rb_hash_lookup2(this_environ, key, Qundef);
+            if (variable != Qundef) {
+                scope = this_environ;
+                goto variable_found;
+            }
+
+            if (!(RTEST(raise_on_not_found) && RTEST(strict_variables))) {
+                // If we aren't running strictly, we need to invoke the default
+                // value proc, rb_hash_aref does this
+                variable = rb_hash_aref(this_environ, key);
+                if (variable != Qnil) {
+                    scope = this_environ;
+                    goto variable_found;
+                }
+            }
```

Comment:
This branch was added in b2118042aa1a5b75849c8753c4f38a39a340da12 entirely to support hashes with default value procs. Ideally instead of `rb_hash_aref` (which looks up the value in the hash again) we could run the default value proc directly, but there is no public C API for this. There is [a test](https://github.com/Shopify/liquid/blob/b0629f17f79a96eb7001b55ea0877b14b53b67a0/test/integration/variable_test.rb#L77-L87) for this in Liquid.